service/awsconfig/v5/resource/cloudformation/update.go

```diff
@@ -64,24 +84,104 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
+	// We enable/disable updates in order to enable them our test installations
+	// but disable them in production installations. That is useful until we have
+	// full confidence in updating guest clusters.
+	if updateallowedcontext.IsUpdateAllowed(ctx) {
+		r.logger.LogCtx(ctx, "debug", "finding out if the main stack has to be updated")
+
+		if shouldUpdate(currentStackState, desiredStackState) {
+			r.logger.LogCtx(ctx, "debug", "main stack has to be updated")
+
+			updateState, err := r.computeUpdateState(customObject, desiredStackState)
+			if err != nil {
+				return cloudformation.CreateStackInput{}, microerror.Mask(err)
+			}
+
+			return updateState, nil
+		} else {
+			r.logger.LogCtx(ctx, "debug", "main stack has not to be updated")
 		}
+	} else {
+		r.logger.LogCtx(ctx, "debug", "not computing update state because main stack are not allowed to be updated")
 	}
 
-	return updateState, nil
+	return cloudformation.UpdateStackInput{}, nil
+}
+
+// shouldScale determines whether the reconciled guest cluster should be scaled.
+// A guest cluster is only allowed to scale in case nothing but the worker count
+// changes. In case anything else changes as well, scaling is not allowed, since
+// any other changes should be covered by general updates, which is a separate
+// step.
+func shouldScale(currentState, desiredState StackState) bool {
+	if currentState.Name != desiredState.Name {
+		return false
+	}
+	if currentState.MasterImageID != desiredState.MasterImageID {
+		return false
+	}
+	if currentState.MasterInstanceType != desiredState.MasterInstanceType {
+		return false
+	}
+	if currentState.MasterCloudConfigVersion != desiredState.MasterCloudConfigVersion {
+		return false
+	}
+	if currentState.WorkerImageID != desiredState.WorkerImageID {
+		return false
+	}
+	if currentState.WorkerInstanceType != desiredState.WorkerInstanceType {
+		return false
+	}
+	if currentState.WorkerCloudConfigVersion != desiredState.WorkerCloudConfigVersion {
+		return false
+	}
+	if currentState.VersionBundleVersion != desiredState.VersionBundleVersion {
+		return false
+	}
+
+	if currentState.WorkerCount != desiredState.WorkerCount {
+		return true
+	}
+
+	return false
+}
+
+// shouldUpdate determines whether the reconciled guest cluster should be
+// updated. A guest cluster is only allowed to update in the following cases.
+//
+//     The update does not indicate scaling.
+//     The version bundle version changes
+//     The instance type of master nodes changes.
+//     The instance type of worker nodes changes.
+//
+func shouldUpdate(currentState, desiredState StackState) bool {
+	if currentState.WorkerCount != desiredState.WorkerCount {
+		return false
+	}
+
+	if currentState.VersionBundleVersion != desiredState.VersionBundleVersion {
+		return true
+	}
+	if currentState.MasterInstanceType != desiredState.MasterInstanceType {
+		return true
+	}
+	if currentState.WorkerInstanceType != desiredState.WorkerInstanceType {
+		return true
+	}
+
+	return false
```

Comment:
Based on my understanding we should only update in case we do not need to scale. Another indicator for updates is a version bundle change and instance type changes which the user can set in the custom object.

Reply:
Good that this now uses the version bundle version. That reduces the number of outputs we need. Good to check the instance types since they come from the custom object.

What happens if the instance type and the worker count are changed at the same time?

That could happen and we can process it in a single stack update. It's useful to separate scaling from updates but I think we can update and scale at the same time.

Reply:
Good shout. I thought about this but was not sure how we want to manage this. We could make it that updates catches updates and scaling and that scaling only catches updates. I would only need to put the updates check in front of the scaling check. When this is fine to you just let me know and I implement this. I can then also align the tests.

Reply:
Yes, let's go with updates catches both updates and scaling. Cloud Formation can handle both and it means we process the updates more efficiently.